File: src/main/java/de/spries/fleetcommander/model/ai/behavior/AggressiveFleetStrategy.java
package de.spries.fleetcommander.model.ai.behavior;

import java.util.List;
import java.util.stream.Collectors;

import de.spries.fleetcommander.model.core.universe.HasCoordinates;
import de.spries.fleetcommander.model.facade.PlayerSpecificPlanet;
import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse;

public class AggressiveFleetStrategy implements FleetStrategy {

	@Override
	public void sendShips(PlayerSpecificUniverse universe) {
		PlayerSpecificPlanet homePlanet = universe.getHomePlanet();
		List<PlayerSpecificPlanet> allPlanets = universe.getPlanets();
		List<PlayerSpecificPlanet> myPlanets = PlayerSpecificPlanet.filterMyPlanets(allPlanets);
		List<PlayerSpecificPlanet> enemyPlanets = allPlanets.stream()
				.filter(p -> p.isKnownAsEnemyPlanet())
				.collect(Collectors.toList());

		if (enemyPlanets.isEmpty()) {
			List<PlayerSpecificPlanet> unknownPlanets = allPlanets.stream()
					.filter(p -> !p.isInhabitedByMe() && p.getIncomingShipCount() == 0)
					.collect(Collectors.toList());
			unknownPlanets = HasCoordinates.sortByDistanceAsc(unknownPlanets, homePlanet);

			int numPlanetsToInvade = Math.min(unknownPlanets.size(), homePlanet.getShipCount());
			List<PlayerSpecificPlanet> planetsToInvade = unknownPlanets.subList(0, numPlanetsToInvade);
			for (PlayerSpecificPlanet unknownPlanet : planetsToInvade) {
				universe.sendShips(1, homePlanet.getId(), unknownPlanet.getId());
			}
		}
		else {
			PlayerSpecificPlanet someEnemyPlanet = enemyPlanets.get(0);
			for (PlayerSpecificPlanet planet : myPlanets) {
				int ships = planet.getShipCount();
				if (ships > 0) {
					universe.sendShips(ships, planet.getId(), someEnemyPlanet.getId());
				}
			}
		}
	}

}

File: src/main/kotlin/de/spries/fleetcommander/model/ai/behavior/AggressiveFleetStrategy.kt
package de.spries.fleetcommander.model.ai.behavior

import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse

open class AggressiveFleetStrategy : FleetStrategy {

    override fun sendShips(universe: PlayerSpecificUniverse) {
        val homePlanet = universe.getHomePlanet()!!
        val allPlanets = universe.getPlanets()
        val myPlanets = allPlanets.filter { p -> p.isInhabitedByMe() }
        val enemyPlanets = allPlanets
                .filter { p -> p.isKnownAsEnemyPlanet() }

        if (enemyPlanets.isEmpty()) {
            val unknownPlanets = allPlanets
                    .filter { p -> !p.isInhabitedByMe() && p.getIncomingShipCount() == 0 }
                    .sortedBy { p -> p.distanceTo(homePlanet) }

            val numPlanetsToInvade = Math.min(unknownPlanets.size, homePlanet.getShipCount())
            val planetsToInvade = unknownPlanets.subList(0, numPlanetsToInvade)
            for (unknownPlanet in planetsToInvade) {
                universe.sendShips(1, homePlanet.getId(), unknownPlanet.getId())
            }
        } else {
            val someEnemyPlanet = enemyPlanets[0]
            for (planet in myPlanets) {
                val ships = planet.getShipCount()
                if (ships > 0) {
                    universe.sendShips(ships, planet.getId(), someEnemyPlanet.getId())
                }
            }
        }
    }
}
